code/ship/ship.cpp

```diff
@@ -13861,92 +13870,104 @@ ship_subsys *ship_get_best_subsys_to_attack(ship *sp, int subsys_type, vec3d *at
-		if ( count == index )
-			return ss;
-		count++;
-		ss = GET_NEXT( ss );
+	// if the indexer already exists, its size won't change, so don't reallocate it
+	if (shipp->subsys_list_indexer.get() == nullptr)
+		shipp->subsys_list_indexer.reset(new ship_subsys* [sinfo->n_subsystems]);
+
+	auto ss = GET_FIRST(&shipp->subsys_list);
+	for (int index = 0; index < sinfo->n_subsystems; ++index)
+	{
+		// normal indexing to valid subsystems
+		if (ss != END_OF_LIST(&shipp->subsys_list))
+		{
+			shipp->subsys_list_indexer.get()[index] = ss;
+			ss->parent_subsys_index = index;
+			ss = GET_NEXT(ss);
+		}
+		// in the event we run out of subsystems (i.e. if not all subsystems were linked)
+		else
+		{
+			shipp->subsys_list_indexer.get()[index] = nullptr;
+		}
 	}
 
-	// get allender -- turret ref didn't fixup correctly!!!!
-	Warning(LOCATION, "In ship_get_indexed_subsys, unable to get a subsystem of index %d on ship %s, due to a broken subsystem reference!  This is most likely due to a table/model mismatch.", index, sp->ship_name);	
-	return NULL;
+	shipp->flags.set(Ship::Ship_Flags::Subsystem_cache_valid);
 }
 
 /**
- * Given a pointer to a subsystem and an associated object, return the index.
+ * Returns the 'nth' ship_subsys structure in a ship's linked list of subsystems.
  */
-int ship_get_index_from_subsys(ship_subsys *ssp, int objnum)
+ship_subsys *ship_get_indexed_subsys(ship *sp, int index)
 {
-	if (ssp == NULL)
-		return -1;
-	else {
-		int	count;
-		ship	*shipp;
-		ship_subsys	*ss;
-
-		Assert(objnum >= 0);
-		Assert(Objects[objnum].instance >= 0);
+	Assertion(index >= 0, "Index must be positive!  The functionality for negative indexes has been moved to ship_get_first_subsys.");
+	Assertion(index < Ship_info[sp->ship_info_index].n_subsystems, "Subsystem index out of range!");
 
-		shipp = &Ships[Objects[objnum].instance];
+	// might need to refresh the cache
+	if (!sp->flags[Ship::Ship_Flags::Subsystem_cache_valid])
+		ship_index_subsystems(sp);
 
-		count = 0;
-		ss = GET_FIRST(&shipp->subsys_list);
-		while ( ss != END_OF_LIST( &shipp->subsys_list ) ) {
-			if ( ss == ssp)
-				return count;
-			count++;
-			ss = GET_NEXT( ss );
-		}
+	return sp->subsys_list_indexer.get()[index];
+}
 
+/**
+* Returns the index number of the ship_subsys parameter within its ship's subsytem list
+*/
+int ship_get_subsys_index(ship_subsys *subsys)
+{
+	if (subsys == nullptr)
 		return -1;
-	}
+
+	// might need to refresh the cache
+	auto sp = &Ships[Objects[subsys->parent_objnum].instance];
+	if (!sp->flags[Ship::Ship_Flags::Subsystem_cache_valid])
+		ship_index_subsystems(sp);
+
+	return subsys->parent_subsys_index;
 }
 
 /**
- * Returns the index number of the ship_subsys parameter
+ * Searches for the subsystem with the given name in the ship's linked list of subsystems, and returns its index or -1 if not found.
  */
-int ship_get_subsys_index(ship *sp, const char* ss_name)
+int ship_get_subsys_index(ship *sp, const char *ss_name)
```

Comment:
This function searches for a C-string.  It's a find-function, so I've renamed it to make that clear.  It can't use the cache because it's not using a pre-computed index; it's performing a search.